### ISSUE TITLE
[sparse]Failed to load cusparse library(cusparse version problem)

### ISSUE THREAD

**Issue by chunleili** [taichi-dev/taichi#8227]
environ: Win10, taichi 1.6.0

**Describe the bug**
My cuda toolkit is 11.8, however driver is 12.1.
So taichi find the wrong version of cusparse(which should be the same as cuda toolkit instead of driver).

**To Reproduce**
```py
import taichi as ti
arch = ti.cuda # or ti.cuda
ti.init(arch=arch)

n = 4
# step 1: create sparse matrix builder
K = ti.linalg.SparseMatrixBuilder(n, n, max_num_triplets=100)

@ti.kernel
def fill(A: ti.types.sparse_matrix_builder()):
    for i in range(n):
        A[i, i] += 1  # Only +=  and -= operators are supported for now.

# step 2: fill the builder with data.
fill(K)

A = K.build()

```

**Log/Screenshots**

```
[W 06/27/23 16:06:43.025 113508] [cuda_driver.cpp:taichi::lang::CUDADriverBase::load_lib@36] cusparse64_12.dll lib not found.
[E 06/27/23 16:06:43.026 113508] [taichi/program/sparse_matrix.h:taichi::lang::CuSparseMatrix::CuSparseMatrix@228] Failed to load cusparse library!


Traceback (most recent call last):
  File "c:\Users\GRBJ200045\Desktop\333.py", line 24, 
in <module>
    A = K.build()
  File "C:\Users\GRBJ200045\.conda\envs\ti-release\lib\site-packages\taichi\linalg\sparse_matrix.py", line 291, in build
    sm = self.ptr.build_cuda()
RuntimeError: [taichi/program/sparse_matrix.h:taichi::lang::CuSparseMatrix::CuSparseMatrix@228] Failed to load cusparse library!
```

**Possible Solutions**
A ad-hoc solution is to rename the `cusparse64_11.dll` to `cusparse64_12.dll`, where my location of this dll is `C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.8\bin`.

A better solution should let the taichi automatically handle the version. And I found the possible location of the code is at https://github.com/taichi-dev/taichi/blob/8be2d6abb815a65006652a48fc9ed0f4bed2e416/taichi/rhi/cuda/cuda_driver.cpp#L131
where the code snippet is
```
  cusparse_loaded_ = load_lib(
      "libcusparse.so",
      "cusparse64_" +
          std::to_string(CUDADriver::get_instance().get_version_major()) +
          ".dll");
```

**Comment by jim19930609** [comment]
The library name of search is determined by:
```
bool CUSPARSEDriver::load_cusparse() {
  cusparse_loaded_ = load_lib(
      "libcusparse.so",
      "cusparse64_" +
          std::to_string(CUDADriver::get_instance().get_version_major()) +
          ".dll");
```

It seems like the CuSparse ".dll" has to be compatible with the CUDA version. So my guess is that you've upgraded your CUDA version but somehow forgot to upgrade the CuSparse library ?

**Comment by houkensjtu** [comment]
Please try out this PR to see if it solves your problem @chunleili

**Comment by houkensjtu** [comment]
Closing the issue since it should be solved by PR #8283

**Comment by reza91ag** [comment]
On the cusparse error: I got rid of it by downloading the whole cuda toolkit after a failed attempt at installing cusparse-dev which is the package cusparse is associated with. Im now running sage-attention with no problem whatsoever. i am not sure if i turned the no-isollation flag on while doing the pip install for sage-attention.
